src/Functions/hostName.cpp

```diff
@@ -1,64 +1,21 @@
-#include <Functions/IFunction.h>
-#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionServerConstantBase.h>
 #include <DataTypes/DataTypeString.h>
 #include <Common/DNSResolver.h>
-#include <Core/Field.h>
-#include <Interpreters/Context.h>
-
 
 namespace DB
 {
 namespace
 {
+    constexpr char name[] = "hostName";
 
-/// Get the host name. Is is constant on single server, but is not constant in distributed queries.
-class FunctionHostName : public IFunction
-{
-public:
-    static constexpr auto name = "hostName";
-    static FunctionPtr create(ContextPtr context)
-    {
-        return std::make_shared<FunctionHostName>(context->isDistributed());
-    }
-
-    explicit FunctionHostName(bool is_distributed_) : is_distributed(is_distributed_)
-    {
-    }
-
-    String getName() const override
-    {
-        return name;
-    }
-
-    bool isDeterministic() const override { return false; }
-
-    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
-
-    bool isDeterministicInScopeOfQuery() const override
-    {
-        return true;
-    }
-
-    bool isSuitableForConstantFolding() const override { return !is_distributed; }
-
-    size_t getNumberOfArguments() const override
-    {
-        return 0;
-    }
-
-    DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
-    {
-        return std::make_shared<DataTypeString>();
-    }
-
-    ColumnPtr executeImpl(const ColumnsWithTypeAndName &, const DataTypePtr & result_type, size_t input_rows_count) const override
+    /// Get the host name. Is is constant on single server, but is not constant in distributed queries.
+    class FunctionHostName : public FunctionServerConstantBase<String, DataTypeString, name>
     {
-        return result_type->createColumnConst(input_rows_count, DNSResolver::instance().getHostName());
-    }
-private:
-    bool is_distributed;
-};
+    public:
+        static FunctionPtr create(ContextPtr context) { return std::make_shared<FunctionHostName>(context); }
```

Comment:
Maybe we can use CRTP pattern to avoid defining this functions in all subclasses?

Reply:
And we can also avoid all template parameters except Derived. So, `T` could be get as `Derived::T`, `ColumnT` as `Derived::ColumnT` and so on...

Reply:
It's possible, but it will look much worse:
```c++

namespace
{
    template <typename Derived>
    struct FunctionServerConstantTraits;
}

/// Base class for functions which return server-level constant like version() or uptime()
template<typename Derived>
class FunctionServerConstantBase : public IFunction
{
public:
    using Value = typename FunctionServerConstantTraits<Derived>::ValueT;
    using Column = typename FunctionServerConstantTraits<Derived>::ColumnT;
    static constexpr auto name = FunctionServerConstantTraits<Derived>::name;
    /// ...
}

namespace
{
    class FunctionZooKeeperSessionUptime;

    template <>
    struct FunctionServerConstantTraits<FunctionZooKeeperSessionUptime>
    {
        static constexpr auto name = "zookeeperSessionUptime";
        using ValueT = UInt32;
        using ColumnT = DataTypeUInt32;
    };

    class FunctionZooKeeperSessionUptime : public FunctionServerConstantBase<FunctionZooKeeperSessionUptime>
    {
    public:
        explicit FunctionZooKeeperSessionUptime(ContextPtr context) : FunctionServerConstantBase(context, context->getZooKeeperSessionUptime()) {}
    };
}
```